File: Override/2.py
class Cal():
    _history = []
    def __init__(self,v1,v2):
        self.v1 = v1
        self.v2 = v2
    def add(self):
        result = self.v1 + self.v2
        Cal._history.append("add : %d + %d = %d" % (self.v1, self.v2, result))
        return result
    def subtract(self):
        result = self.v1 - self.v2
        Cal._history.append("subtract : %d - %d = %d" % (self.v1, self.v2, result))
        return result
    @classmethod
    def history(cls):
        for item in Cal._history:
            print(item)

class CalMultiply(Cal):
    def multiply(self):
        result = self.v1 * self.v2
        Cal._history.append("multiply: %d * %d = %d" % (self.v1, self.v2, result))
        return result

class CalDivide(CalMultiply):
    def divide(self):
        result = self.v1 / self.v2
        Cal._history.append("divide: %d / %d = %d" % (self.v1, self.v2, result))
        return result

c1 = Cal(20,10)
print(c1.add())
print(c1.subtract())

c2 = CalMultiply(20,10)
print(c2.multiply())
print(c2.add())

c3 = CalDivide(20,10)
print(c3.divide())
Cal.history()

File: Override/2.rb
class Cal
    @@_history = []
    def initialize(v1, v2)
      @v1 = v1
      @v2 = v2
    end
    def add()
      result = @v1 + @v2
      @@_history.push("add : #{@v1} + #{@v2} = #{result} " )
      return result
    end
    def subtract()
      result = @v1 - @v2
      @@_history.push("subtract : #{@v1} - #{@v2} = #{result} " )
      return result
    end
    def Cal.history()
      for item in @@_history
      p item
      end
    end
end

class CalMultiply < Cal
  def multiply()
    result = @v1 * @v2
    @@_history.push("multiply : #{@v1} * #{@v2} = #{result} " )
    return result
  end
end

class CalDivide < CalMultiply
  def divide()
    result = @v1 / @v2
    @@_history.push("divide : #{@v1} / #{@v2} = #{result} " )
    return result
  end
end


c1 = Cal.new(20,10)
p c1, c1.add

c3 = CalDivide.new(20,10)
p c3, c3.divide
p c3.add
p Cal.history
